### ISSUE TITLE
how do you train the autoregressive predictor

### ISSUE THREAD

**Issue by CuriousCat-7** [SihanXU/nepa#8]
<img width="1430" height="1096" alt="Image" src="https://github.com/user-attachments/assets/9e50528b-cab7-413f-9ade-3f5348db7c43" />

I'm not familiar with the representation-learning area, 
Could you please tell me that if the gradient is stopped, there's no grad from input_embd to pred_embd, then how to train the h.param?

thank you

**Comment by SihanXU** [comment]
Detaching `z` only stops gradients through the **target** branch. The loss still depends on `z_hat` (the prediction), so gradients flow `loss → z_hat → h.param` (and also `loss → z_hat → input_embed → f.param`). Hence `h` is trained normally.

**Comment by CuriousCat-7** [comment]
> Detaching `z` only stops gradients through the **target** branch. The loss still depends on `z_hat` (the prediction), so gradients flow `loss → z_hat → h.param` (and also `loss → z_hat → input_embed → f.param`). Hence `h` is trained normally.

Oh, I see! I accidentally confused `z.detach()` with `z_hat.detach()`.
